<launch>

      <arg name="publish_clock" default="--clock"/>
      <arg name="autorun" default="0"/>
      <arg name="loop_en" default="0"/>
      <arg name="show_report"   default="0"/>

      <!-- MCDVIRAL dataset, downloadable at mcdviral.github.io. -->
      
      <arg name="data_path" default="/home/dat/data/megaparticle"/>


      <arg  name="bag_file"  default="$(arg data_path)/outdoor_kidnap_a/*.bag"/>
      <!-- <arg  name="bag_file"  default="$(arg data_path)/outdoor_kidnap_b/*.bag"/> -->

      <!-- <arg  name="bag_file"  default="$(arg data_path)/outdoor_hard_01a/*.bag"/> -->
      <!-- <arg  name="bag_file"  default="$(arg data_path)/outdoor_hard_01b/*.bag"/> -->

      <!-- <arg  name="bag_file"  default="$(arg data_path)/outdoor_hard_02a/*.bag"/> -->
      <!-- <arg  name="bag_file"  default="$(arg data_path)/outdoor_hard_02b/*.bag"/> -->

      <!-- Enable prior map -->
      <arg name="use_prior_map" default="1"/>

      <!-- Manual initialization for reloc -->
      <!-- <group if="$(eval 'outdoor_kidnap_a' in bag_file)"> <rosparam param="/tf_Lprior_L0_init"> [56.14, 106.63, 7.58,  18, 0, 0] </rosparam> </group>
      <group if="$(eval 'outdoor_kidnap_b' in bag_file)"> <rosparam param="/tf_Lprior_L0_init"> [55.07, 105.76, 7.71, -17, 0, 0] </rosparam> </group> -->

      <!-- Chose the config file based on the sequence names -->
      <arg name="config_file" value="$(find slict)/config/megaparticle_outdoor.yaml"/>

      <!-- Priormap file -->
      <arg name="prior_map_dir" default="/home/dat/data/megaparticle/priormap/outdoor_kidnap"/>
      <!-- <arg name="prior_map_dir" default="/home/dat/data/megaparticle/priormap/outdoor_hard"/> -->

      <!-- Directory to log the bundle adjustment poses, pointclouds, and others-->
      <arg name="exp_log_dir" default="/home/$(env USER)/slict_logs/megaparticle_outdoor"/>

      <!-- Load the param -->
      <rosparam file="$(arg config_file)" command="load"/>
      <!-- Set some exp related params -->
      <param name="/autoexit" type="int"    value="$(arg autorun)" />
      <param name="/loop_en"  type="int"    value="$(arg loop_en)" />
      <param name="/log_dir"  type="string" value="$(arg exp_log_dir)" />

      <!-- Run the livox to ouster converter -->
      <node pkg="slict" required="true" type="slict_livox2_to_ouster" name="slict_livox2_to_ouster" respawn="false" output="screen"/>

      <!-- Run the sensor sync -->
      <node pkg="slict" required="true" type="slict_sensorsync" name="slict_sensorsync" respawn="false" output="log"/>

      <!-- Run the relocalization -->
      <!-- <node pkg="slict" type="slict_relocalization" name="slict_relocalization" respawn="false" output="screen" > -->
      <node pkg="slict" type="slict_relocalization" name="slict_relocalization" respawn="false" >
        <param name="/priormap_dir"  type="string" value="$(arg prior_map_dir)" />
        <rosparam file="$(find slict)/config/multi_reloc_megaparticle_outdoor.yaml" command="load" />
        <remap from="/liteloam_pose"  to="/reloc_pose"/>
      </node>


      <!-- <node pkg="std_detector"
            type="relocalization_ntu_online"
            name="relocalization_ntu">

      <rosparam command="load" file="$(find std_detector)/config/config_megaparticle_outdoor.yaml" />
      <param name="descriptor_file_path"         type="string" value = "/home/dat/data/std_reloc/megaparticle/saved_descriptors_outdoor_kidnap/"/>

      <remap from="/relocalization_pose" to="/uwb_pose"/>
      </node> -->


      <!-- Run the Estimator -->
      <param name="/prior_map_dir" type="string" value="$(arg prior_map_dir)" />
      <param name="/use_prior_map" type="int"    value="$(arg use_prior_map)" />
      <param name="/show_report"    type="int"    value="$(arg show_report)"   />
      <param name="/slam_ref_frame" type="string" value="odom"/>
      <node pkg="slict" required="true" type="slict_estimator" name="slict_estimator" respawn="false" output="screen"/>
      
      <!--- Run Rviz-->
      <node pkg="rviz" type="rviz" name="rviz" respawn="true" output="log"
            args="-d $(find slict)/launch/megaparticle_outdoor.rviz" />

      <!-- Play the bag file -->
      <param name="/use_sim_time" value="false" />
      <arg name="wait_time" value="$(eval use_prior_map*5 + 1)" />
      <node required="false" pkg="rosbag" type="play" name="rosbag_player"
            args="$(arg publish_clock) $(arg bag_file) -r 1.0"
            launch-prefix="bash -c 'sleep $(arg wait_time); $0 $@' "/>

      <!-- <node pkg="tf" type="static_transform_publisher" name="base2view" args="0 0 0 0 0 0 world local 1"/> -->
      <node pkg="slict" type="odom_to_tf.py" name="odom_to_tf" output="screen" respawn="false"/>

      <!-- <node pkg="slict" type="reloc_pose_to_csv.py" name="reloc_pose_to_csv" output="screen" respawn="false" >    
            <param name="output_dir" value="/home/dat/data/exp/outdoor_kidnap/reloc_pose" />
            <param name="csv_name" value="outdoor_kidnap_a_reloc.csv" />
      </node> -->

</launch>